src/python/pants/core/goals/typecheck.py

```diff
@@ -170,6 +184,33 @@ async def typecheck(
         Get(EnrichedTypecheckResults, TypecheckRequest, request) for request in valid_requests
     )
 
+    # Handle reports.
+    disambiguated_dirs: set[str] = set()
+
+    def write_report(digest: Digest, subdir: str) -> None:
```

Comment:
Could probably break this out into a helper since it's now copy-pasted in two places?

Reply:
Done, which faciliated adding a test for all these edge cases.